File: src/alert_history/core/interfaces.py
"""
Core interfaces for Alert History Service.

Implements SOLID principles through well-defined interfaces:
- Single Responsibility: Each interface has one clear purpose
- Interface Segregation: Small, focused interfaces
- Dependency Inversion: Depend on abstractions, not concretions
"""

# Standard library imports
from abc import ABC, abstractmethod
from dataclasses import dataclass
from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional


class AlertSeverity(Enum):
    """Alert severity levels."""

    CRITICAL = "critical"
    WARNING = "warning"
    INFO = "info"
    NOISE = "noise"


class AlertStatus(Enum):
    """Alert status values."""

    FIRING = "firing"
    RESOLVED = "resolved"


class PublishingFormat(Enum):
    """Publishing format options."""

    ALERTMANAGER = "alertmanager"
    ROOTLY = "rootly"
    PAGERDUTY = "pagerduty"
    SLACK = "slack"
    WEBHOOK = "webhook"


@dataclass
class Alert:
    """Alert data model."""

    fingerprint: str
    alert_name: str
    status: AlertStatus
    labels: Dict[str, str]
    annotations: Dict[str, str]
    starts_at: datetime
    ends_at: Optional[datetime] = None
    generator_url: Optional[str] = None

    @property
    def namespace(self) -> Optional[str]:
        """Get alert namespace from labels."""
        return self.labels.get("namespace")

    @property
    def severity(self) -> Optional[str]:
        """Get alert severity from labels."""
        return self.labels.get("severity")


@dataclass
class ClassificationResult:
    """LLM classification result."""

    severity: AlertSeverity
    confidence: float
    reasoning: str
    recommendations: List[str]
    processing_time: float
    metadata: Optional[Dict[str, Any]] = None


@dataclass
class PublishingTarget:
    """Publishing target configuration."""

    name: str
    type: str
    url: str
    enabled: bool
    filter_config: Dict[str, Any]
    headers: Dict[str, str]
    format: PublishingFormat


@dataclass
class EnrichedAlert:
    """Alert enriched with classification data."""

    alert: Alert
    classification: Optional[ClassificationResult] = None
    enrichment_metadata: Optional[Dict[str, Any]] = None
    processing_timestamp: Optional[datetime] = None


# Storage Interfaces (Single Responsibility Principle)


class IAlertStorage(ABC):
    """Interface for alert storage operations."""

    @abstractmethod
    async def save_alert(self, alert: Alert) -> bool:
        """Save alert to storage."""
        pass

    @abstractmethod
    async def get_alert_by_fingerprint(self, fingerprint: str) -> Optional[Alert]:
        """Get alert by fingerprint."""
        pass

    @abstractmethod
    async def get_alerts(
        self, filters: Dict[str, Any], limit: int = 100, offset: int = 0
    ) -> List[Alert]:
        """Get alerts with filters."""
        pass

    @abstractmethod
    async def cleanup_old_alerts(self, retention_days: int) -> int:
        """Clean up old alerts and return count of deleted records."""
        pass


class IClassificationStorage(ABC):
    """Interface for classification storage operations."""

    @abstractmethod
    async def save_classification(
        self, fingerprint: str, result: ClassificationResult
    ) -> bool:
        """Save classification result."""
        pass

    @abstractmethod
    async def get_classification(
        self, fingerprint: str
    ) -> Optional[ClassificationResult]:
        """Get classification by fingerprint."""
        pass


class IPublishingLogStorage(ABC):
    """Interface for publishing log storage."""

    @abstractmethod
    async def log_publishing_attempt(
        self,
        fingerprint: str,
        target_name: str,
        success: bool,
        error_message: Optional[str] = None,
        processing_time: Optional[float] = None,
    ) -> bool:
        """Log publishing attempt."""
        pass


# Caching Interfaces (Interface Segregation Principle)


class ICache(ABC):
    """Generic cache interface."""

    @abstractmethod
    async def get(self, key: str) -> Optional[Any]:
        """Get value from cache."""
        pass

    @abstractmethod
    async def set(self, key: str, value: Any, ttl: Optional[int] = None) -> bool:
        """Set value in cache."""
        pass

    @abstractmethod
    async def delete(self, key: str) -> bool:
        """Delete value from cache."""
        pass

    @abstractmethod
    async def exists(self, key: str) -> bool:
        """Check if key exists in cache."""
        pass


class IDistributedLock(ABC):
    """Interface for distributed locking."""

    @abstractmethod
    async def acquire_lock(self, key: str, timeout: int = 60) -> bool:
        """Acquire distributed lock."""
        pass

    @abstractmethod
    async def release_lock(self, key: str) -> bool:
        """Release distributed lock."""
        pass


# LLM Service Interfaces (Single Responsibility Principle)


class ILLMClient(ABC):
    """Interface for LLM communication."""

    @abstractmethod
    async def classify_alert(
        self, alert: Alert, context: Optional[Dict[str, Any]] = None
    ) -> ClassificationResult:
        """Classify alert using LLM."""
        pass

    @abstractmethod
    async def generate_recommendations(
        self, alert: Alert, classification: ClassificationResult
    ) -> List[str]:
        """Generate configuration recommendations."""
        pass


class IAlertClassifier(ABC):
    """Interface for alert classification service."""

    @abstractmethod
    async def classify(self, alert: Alert) -> ClassificationResult:
        """Classify alert with caching and error handling."""
        pass


# Publishing Interfaces (Interface Segregation Principle)


class IAlertFormatter(ABC):
    """Interface for alert formatting."""

    @abstractmethod
    async def format_alert(
        self, enriched_alert: EnrichedAlert, target_format: PublishingFormat
    ) -> Dict[str, Any]:
        """Format alert for specific target."""
        pass


class IAlertPublisher(ABC):
    """Interface for alert publishing."""

    @abstractmethod
    async def publish_alert(
        self, enriched_alert: EnrichedAlert, target: PublishingTarget
    ) -> bool:
        """Publish alert to target."""
        pass


class IFilterEngine(ABC):
    """Interface for alert filtering."""

    @abstractmethod
    async def should_publish(
        self, enriched_alert: EnrichedAlert, target: PublishingTarget
    ) -> bool:
        """Check if alert should be published to target."""
        pass


# Configuration Management Interfaces (Dependency Inversion Principle)


class IConfigurationManager(ABC):
    """Interface for configuration management."""

    @abstractmethod
    async def get_config(self, key: str, default: Any = None) -> Any:
        """Get configuration value."""
        pass

    @abstractmethod
    async def get_all_configs(self) -> Dict[str, Any]:
        """Get all configuration values."""
        pass

    @abstractmethod
    async def reload_config(self) -> bool:
        """Reload configuration from source."""
        pass


class ISecretsManager(ABC):
    """Interface for secrets management."""

    @abstractmethod
    async def get_secret(self, key: str) -> Optional[str]:
        """Get secret value."""
        pass

    @abstractmethod
    async def list_secrets(self, label_selector: str) -> Dict[str, Dict[str, str]]:
        """List secrets matching label selector."""
        pass


# Target Discovery Interface (Open/Closed Principle)


class ITargetDiscovery(ABC):
    """Interface for dynamic target discovery."""

    @abstractmethod
    async def discover_targets(self) -> List[PublishingTarget]:
        """Discover available publishing targets."""
        pass

    @abstractmethod
    async def refresh_targets(self) -> bool:
        """Refresh target configuration."""
        pass


# Health Check Interface (Single Responsibility Principle)


class IHealthChecker(ABC):
    """Interface for health checking."""

    @abstractmethod
    async def check_health(self) -> Dict[str, Any]:
        """Perform health check."""
        pass

    @abstractmethod
    async def check_readiness(self) -> Dict[str, Any]:
        """Perform readiness check."""
        pass


# Metrics Interface (Interface Segregation Principle)


class IMetricsCollector(ABC):
    """Interface for metrics collection."""

    @abstractmethod
    def increment_counter(
        self, name: str, labels: Optional[Dict[str, str]] = None
    ) -> None:
        """Increment counter metric."""
        pass

    @abstractmethod
    def set_gauge(
        self, name: str, value: float, labels: Optional[Dict[str, str]] = None
    ) -> None:
        """Set gauge metric."""
        pass

    @abstractmethod
    def observe_histogram(
        self, name: str, value: float, labels: Optional[Dict[str, str]] = None
    ) -> None:
        """Observe histogram metric."""
        pass


# Event Processing Interface (Strategy Pattern for extensibility)


class IEventProcessor(ABC):
    """Interface for event processing strategies."""

    @abstractmethod
    async def process_event(self, event_data: Dict[str, Any]) -> bool:
        """Process incoming event."""
        pass

    @abstractmethod
    def can_handle(self, event_type: str) -> bool:
        """Check if processor can handle event type."""
        pass


# Repository Pattern Interfaces (DRY principle)


class IRepository(ABC):
    """Generic repository interface."""

    @abstractmethod
    async def create(self, entity: Any) -> bool:
        """Create entity."""
        pass

    @abstractmethod
    async def get_by_id(self, entity_id: str) -> Optional[Any]:
        """Get entity by ID."""
        pass

    @abstractmethod
    async def update(self, entity: Any) -> bool:
        """Update entity."""
        pass

    @abstractmethod
    async def delete(self, entity_id: str) -> bool:
        """Delete entity."""
        pass

    @abstractmethod
    async def list(
        self,
        filters: Optional[Dict[str, Any]] = None,
        limit: int = 100,
        offset: int = 0,
    ) -> List[Any]:
        """List entities with optional filters."""
        pass

File: go-app/internal/core/interfaces.go
package core

import (
	"context"
	"time"
)

// AlertSeverity represents alert severity levels
type AlertSeverity string

const (
	SeverityCritical AlertSeverity = "critical"
	SeverityWarning  AlertSeverity = "warning"
	SeverityInfo     AlertSeverity = "info"
	SeverityNoise    AlertSeverity = "noise"
)

// AlertStatus represents alert status values
type AlertStatus string

const (
	StatusFiring   AlertStatus = "firing"
	StatusResolved AlertStatus = "resolved"
)

// PublishingFormat represents publishing format options
type PublishingFormat string

const (
	FormatAlertmanager PublishingFormat = "alertmanager"
	FormatRootly       PublishingFormat = "rootly"
	FormatPagerDuty    PublishingFormat = "pagerduty"
	FormatSlack        PublishingFormat = "slack"
	FormatWebhook      PublishingFormat = "webhook"
)

// Alert represents alert data model
type Alert struct {
	Fingerprint  string            `json:"fingerprint" validate:"required"`
	AlertName    string            `json:"alert_name" validate:"required"`
	Status       AlertStatus       `json:"status" validate:"required,oneof=firing resolved"`
	Labels       map[string]string `json:"labels"`
	Annotations  map[string]string `json:"annotations"`
	StartsAt     time.Time         `json:"starts_at" validate:"required"`
	EndsAt       *time.Time        `json:"ends_at,omitempty"`
	GeneratorURL *string           `json:"generator_url,omitempty" validate:"omitempty,url"`
	Timestamp    *time.Time        `json:"timestamp,omitempty"`
}

// Namespace returns alert namespace from labels
func (a *Alert) Namespace() *string {
	if ns, ok := a.Labels["namespace"]; ok {
		return &ns
	}
	return nil
}

// Severity returns alert severity from labels
func (a *Alert) Severity() *string {
	if sev, ok := a.Labels["severity"]; ok {
		return &sev
	}
	return nil
}

// ClassificationResult represents LLM classification result
type ClassificationResult struct {
	Severity        AlertSeverity  `json:"severity" validate:"required,oneof=critical warning info noise"`
	Confidence      float64        `json:"confidence" validate:"gte=0,lte=1"`
	Reasoning       string         `json:"reasoning" validate:"required"`
	Recommendations []string       `json:"recommendations"`
	ProcessingTime  float64        `json:"processing_time" validate:"gte=0"`
	Metadata        map[string]any `json:"metadata,omitempty"`
}

// PublishingTarget represents publishing target configuration
type PublishingTarget struct {
	Name         string            `json:"name" validate:"required"`
	Type         string            `json:"type" validate:"required"`
	URL          string            `json:"url" validate:"required,url"`
	Enabled      bool              `json:"enabled"`
	FilterConfig map[string]any    `json:"filter_config"`
	Headers      map[string]string `json:"headers"`
	Format       PublishingFormat  `json:"format" validate:"required,oneof=alertmanager rootly pagerduty slack webhook"`
}

// EnrichedAlert represents alert enriched with classification data
type EnrichedAlert struct {
	Alert               *Alert                `json:"alert"`
	Classification      *ClassificationResult `json:"classification,omitempty"`
	EnrichmentMetadata  map[string]any        `json:"enrichment_metadata,omitempty"`
	ProcessingTimestamp *time.Time            `json:"processing_timestamp,omitempty"`
}

// Database interfaces following SOLID principles

// TimeRange represents time range filter
type TimeRange struct {
	From *time.Time `json:"from,omitempty"`
	To   *time.Time `json:"to,omitempty"`
}

// AlertFilters represents filters for alert queries
type AlertFilters struct {
	Status    *AlertStatus      `json:"status,omitempty"`
	Severity  *string           `json:"severity,omitempty"`
	Namespace *string           `json:"namespace,omitempty"`
	Labels    map[string]string `json:"labels,omitempty"`
	TimeRange *TimeRange        `json:"time_range,omitempty"`
	Limit     int               `json:"limit" validate:"gte=0,lte=1000"`
	Offset    int               `json:"offset" validate:"gte=0"`
}

// Validate validates AlertFilters parameters
func (f *AlertFilters) Validate() error {
	// Validate Limit
	if f.Limit < 0 {
		return ErrInvalidFilterLimit
	}
	if f.Limit > 1000 {
		return ErrFilterLimitTooLarge
	}

	// Validate Offset
	if f.Offset < 0 {
		return ErrInvalidFilterOffset
	}

	// Validate Status
	if f.Status != nil {
		if *f.Status != StatusFiring && *f.Status != StatusResolved {
			return ErrInvalidFilterStatus
		}
	}

	// Validate Severity
	if f.Severity != nil {
		validSeverities := map[string]bool{
			"critical": true,
			"warning":  true,
			"info":     true,
			"noise":    true,
		}
		if !validSeverities[*f.Severity] {
			return ErrInvalidFilterSeverity
		}
	}

	// Validate TimeRange
	if f.TimeRange != nil {
		if f.TimeRange.From != nil && f.TimeRange.To != nil {
			if f.TimeRange.From.After(*f.TimeRange.To) {
				return ErrInvalidTimeRange
			}
		}
	}

	// Validate Labels (max 20 labels)
	if len(f.Labels) > 20 {
		return ErrTooManyLabels
	}

	// Validate label keys and values (max 255 chars each)
	for key, value := range f.Labels {
		if len(key) == 0 {
			return ErrEmptyLabelKey
		}
		if len(key) > 255 {
			return ErrLabelKeyTooLong
		}
		if len(value) > 255 {
			return ErrLabelValueTooLong
		}
	}

	return nil
}

// AlertList represents paginated list of alerts
type AlertList struct {
	Alerts []*Alert `json:"alerts"`
	Total  int      `json:"total"`
	Limit  int      `json:"limit"`
	Offset int      `json:"offset"`
}

// AlertStats represents alert statistics
type AlertStats struct {
	TotalAlerts       int            `json:"total_alerts"`
	AlertsByStatus    map[string]int `json:"alerts_by_status"`
	AlertsBySeverity  map[string]int `json:"alerts_by_severity"`
	AlertsByNamespace map[string]int `json:"alerts_by_namespace"`
	OldestAlert       *time.Time     `json:"oldest_alert,omitempty"`
	NewestAlert       *time.Time     `json:"newest_alert,omitempty"`
}

// AlertStorage interface for alert storage operations
type AlertStorage interface {
	// Basic CRUD operations
	SaveAlert(ctx context.Context, alert *Alert) error
	GetAlertByFingerprint(ctx context.Context, fingerprint string) (*Alert, error)
	ListAlerts(ctx context.Context, filters *AlertFilters) (*AlertList, error)
	UpdateAlert(ctx context.Context, alert *Alert) error
	DeleteAlert(ctx context.Context, fingerprint string) error

	// Additional operations
	GetAlertStats(ctx context.Context) (*AlertStats, error)
	CleanupOldAlerts(ctx context.Context, retentionDays int) (int, error)
}

// ClassificationStorage interface for classification storage operations
type ClassificationStorage interface {
	SaveClassification(ctx context.Context, fingerprint string, result *ClassificationResult) error
	GetClassification(ctx context.Context, fingerprint string) (*ClassificationResult, error)
}

// PublishingLogStorage interface for publishing log storage
type PublishingLogStorage interface {
	LogPublishingAttempt(ctx context.Context, fingerprint, targetName string, success bool, errorMessage *string, processingTime *float64) error
	GetPublishingHistory(ctx context.Context, fingerprint string) ([]*PublishingLog, error)
}

// PublishingLog represents publishing attempt log
type PublishingLog struct {
	ID             string    `json:"id"`
	Fingerprint    string    `json:"fingerprint"`
	TargetName     string    `json:"target_name"`
	Success        bool      `json:"success"`
	ErrorMessage   *string   `json:"error_message,omitempty"`
	ProcessingTime *float64  `json:"processing_time,omitempty"`
	CreatedAt      time.Time `json:"created_at"`
}

// Combined Database interface for full functionality
type Database interface {
	// Core database operations
	Connect(ctx context.Context) error
	Disconnect(ctx context.Context) error
	Health(ctx context.Context) error

	// Alert operations
	AlertStorage

	// Classification operations
	ClassificationStorage

	// Publishing operations
	PublishingLogStorage

	// Migration operations
	MigrateUp(ctx context.Context) error
	MigrateDown(ctx context.Context, steps int) error

	// Utility operations
	GetStats(ctx context.Context) (map[string]interface{}, error)
}

// Cache interfaces

// Cache interface for generic caching
type Cache interface {
	Get(ctx context.Context, key string) (any, error)
	Set(ctx context.Context, key string, value any, ttl *time.Duration) error
	Delete(ctx context.Context, key string) error
	Exists(ctx context.Context, key string) (bool, error)
}

// DistributedLock interface for distributed locking
type DistributedLock interface {
	AcquireLock(ctx context.Context, key string, timeout time.Duration) (bool, error)
	ReleaseLock(ctx context.Context, key string) error
}

// LLM Service interfaces

// LLMClient interface for LLM communication
type LLMClient interface {
	ClassifyAlert(ctx context.Context, alert *Alert, context map[string]any) (*ClassificationResult, error)
	GenerateRecommendations(ctx context.Context, alert *Alert, classification *ClassificationResult) ([]string, error)
}

// AlertClassifier interface for alert classification service
type AlertClassifier interface {
	Classify(ctx context.Context, alert *Alert) (*ClassificationResult, error)
}

// Publishing interfaces

// AlertFormatter interface for alert formatting
type AlertFormatter interface {
	FormatAlert(ctx context.Context, enrichedAlert *EnrichedAlert, targetFormat PublishingFormat) (map[string]any, error)
}

// AlertPublisher interface for alert publishing
type AlertPublisher interface {
	PublishAlert(ctx context.Context, enrichedAlert *EnrichedAlert, target *PublishingTarget) error
}

// FilterEngine interface for alert filtering
type FilterEngine interface {
	ShouldPublish(ctx context.Context, enrichedAlert *EnrichedAlert, target *PublishingTarget) (bool, error)
}

// Configuration Management interfaces

// ConfigurationManager interface for configuration management
type ConfigurationManager interface {
	GetConfig(ctx context.Context, key string, defaultValue any) (any, error)
	GetAllConfigs(ctx context.Context) (map[string]any, error)
	ReloadConfig(ctx context.Context) error
}

// SecretsManager interface for secrets management
type SecretsManager interface {
	GetSecret(ctx context.Context, key string) (string, error)
	ListSecrets(ctx context.Context, labelSelector string) (map[string]map[string]string, error)
}

// TargetDiscovery interface for dynamic target discovery
type TargetDiscovery interface {
	DiscoverTargets(ctx context.Context) ([]*PublishingTarget, error)
	RefreshTargets(ctx context.Context) error
}

// Health Check interface

// HealthChecker interface for health checking
type HealthChecker interface {
	CheckHealth(ctx context.Context) (map[string]any, error)
	CheckReadiness(ctx context.Context) (map[string]any, error)
}

// Metrics interface

// MetricsCollector interface for metrics collection
type MetricsCollector interface {
	IncrementCounter(ctx context.Context, name string, labels map[string]string)
	SetGauge(ctx context.Context, name string, value float64, labels map[string]string)
	ObserveHistogram(ctx context.Context, name string, value float64, labels map[string]string)
}

// Event Processing interface

// EventProcessor interface for event processing strategies
type EventProcessor interface {
	ProcessEvent(ctx context.Context, eventData map[string]any) error
	CanHandle(eventType string) bool
}

// Repository Pattern interface

// Repository interface for generic repository operations
type Repository[T any] interface {
	Create(ctx context.Context, entity *T) error
	GetByID(ctx context.Context, id string) (*T, error)
	Update(ctx context.Context, entity *T) error
	Delete(ctx context.Context, id string) error
	List(ctx context.Context, filters map[string]any, limit, offset int) ([]*T, error)
}
